### PR TITLE
Branch validation: Add HA validation

### PR DESCRIPTION
**Pull request by angstwad**

### CODE REVIEW

File: branch-validation.yml [hunk 1]
```diff
@@ -1,6 +1,10 @@
-
-# playbook for validating a branch for release tagging
+# Playbook for validating a branch for release tagging
 # interactive, may prompt runner to affirm things have completed
+
+# By default, this launches auto-sechedueld instances.  In order to launch
```
Comment (j2sol) [inline branch-validation.yml]: Typo.

**Comment by j2sol** [comment]
Feedback given in Slack. An idea to make this not rely on setting ha_group and using discovered group membership to decide where to boot instances.

File: roles/branch-validation/tasks/main.yml [hunk 2]
```diff
@@ -71,6 +71,28 @@
     wait: yes
   with_items: groups[instance_group]
   register: instances
+  when: not groups['compute'] | intersect(groups['controller']) | length | bool
```
Comment (j2sol) [inline roles/branch-validation/tasks/main.yml]: This role is running on `controller[0]`, so we already know we're in the controller group. Thus all we should have to check for is if we're ALSO in the compute group, so you could do `when: not 'compute' in groups`.

However, I'm even more unclear what the approach is here, why there are two different instance launching tasks. Why can't they be combined into one and just iterate over the hosts in groups['compute'] ?

Comment (angstwad) [inline roles/branch-validation/tasks/main.yml]: So, it runs from `controller[0]` but it attempts to schedule instances on each AZ, and this, I think, really all depends on whether or not it can schedule to a control node.  So, AFAICT, when I read the original playbook, it was trying to a) launch an instance on each compute node and b) if there were converged controllers, schedule on each converged control node explicitly.

I made some guesses based on my interpretation of the intent of the original playbook so this could be...wrong.  IDK.

File: roles/branch-validation/tasks/main.yml [hunk 3]
```diff
@@ -113,6 +125,7 @@
   delay: 1
   retries: 5
   with_items: instances.results
+  when: not groups['compute'] | intersect(groups['controller']) | length | bool
```
Comment (j2sol) [inline roles/branch-validation/tasks/main.yml]: Why is this conditional here? The only things inside of `instances.results` would be the instances we spun up, one for every computer in the `instance_group` group (which, we should go back to making that hard set to `compute` since there isn't a way in Ursula for this group to be anything different).

File: roles/branch-validation/tasks/main.yml [hunk 4]
```diff
@@ -155,7 +168,7 @@
     that: ping.user_input|bool
 
 - include: ha.yml
-  when: validate_ha_environment
+  when: groups['compute'] | intersect(groups['controller']) | length | bool
```
Comment (j2sol) [inline roles/branch-validation/tasks/main.yml]: This conditional here would prevent all of the HA testing from happening when on dedicated controller environments. That's not what we want.

We still want to test that HA works, it's just that we can't check if instances that were on controllers come back after a controller reboot. Those are the tasks that we need to add a condition to, and adjust the other tasks within that assume there being instances on controllers and computes.

**Comment by j2sol** [comment]
retest

File: roles/nova-data/tasks/ssh.yml [hunk 5]
```diff
@@ -38,3 +38,8 @@
 
 - name: verify ssh among compute nodes
   command: sudo -u nova -H /bin/sh -c {{ nova.state_path }}/bin/verify-ssh
+
+- name: Install nova ssh keys for root
+  authorized_keys:
+    key: "{{ hostvars[item].nova_user.ssh_public_key }}"
+  with_items: groups['compute']
```
Comment (j2sol) [inline roles/nova-data/tasks/ssh.yml]: This change seems unrelated, probably from your other branch.

File: roles/branch-validation/tasks/main.yml [hunk 6]
```diff
@@ -155,7 +167,7 @@
     that: ping.user_input|bool
 
 - include: ha.yml
-  when: validate_ha_environment
+  when: not groups['compute'] | intersect(groups['controller']) | length | bool
```
Comment (j2sol) [inline roles/branch-validation/tasks/main.yml]: Can you help me understand what this conditional is trying to accomplish?

We want to validate ha, regardless if the environment is hyper converged (compute on controllers) or not.

What we need to do is adjust the tasks inside ha.yml so that there aren't any assumptions made about where the booted instances are.

Comment (angstwad) [inline roles/branch-validation/tasks/main.yml]: the point of this is for one thing:  when an environment has converged controllers, compute nodes that are also controllers, then we'll run the HA tests.

This is possible because instances are launched (one each for each compute note) and scheduled on each respective compute node.  This validates scheduling and ensures that if an environment has converged controllers, that we're not missing them.  

So, if the intersection of `controller` and `compute` has a non-zero length, we have converged controllers, and we should test them by running ha.yml.

Comment (j2sol) [inline roles/branch-validation/tasks/main.yml]: So here's my problem. We need to test HA whether or not the cluster has converged controllers. We still have to test that having one controller go down does not disrupt the data path for the instances in the cluster.

It's just that if the cluster is not converged, we shouldn't worry about the status of instances on the downed controller coming back, because there are no instances on the downed controller.

Does that make sense?
